Semantic diffings for functor types and applications

Drup opened the pull request:
This is a patch made by @Octachron and myself to improve the error messages for ill-typed functor applications and inclusion between functor types. It does so by computing a diff between the expected and the provided types, and find an application that minimizes the error. The set of changes also paves the way for better diff for inclusion checks between signatures.

As a first, recurring example, it detects missing arguments:
```ocaml
module X: x = struct type a end
module Y: y = struct type b end
module Z: z = struct type c end
module F(X:x)(Y:y)(Z:z) = struct end
# module M = F(X)(Z) ;;
             ^^^^^^^
Error: The functor application is ill-typed.
       These arguments:
         X  Z
       do not match these parameters:
         functor (X : x) (Y : y) (Z : z) -> ...
  1. Module X matches the expected module type x
  2. An argument appears to be missing with module type y
  3. Module Z matches the expected module type z
```

Each "bullet point" represent a position in the arguments, and the associated errors (or lack thereof).

It scales to fairly sophisticated (i.e. : "which never happens in reality") errors:
```ocaml
# module type ARG = sig
    module type A = sig type a end
    module type Foo
    module X:   Foo
    module Y:   A
    module G(Z:sig type t end)(Y:sig type a end)(_:A)(Z:A) : sig end
    type u = G(X)(Y)(X)(Y)(X).t
  end ;;
Error: The functor application G(X)(Y)(X)(Y)(X) is ill-typed.
       These arguments:
         X Y X Y X
       do not match these parameters:
         functor (Z : ...(Z)) (Y : ...) A (Z : A)  -> ...
  1. Modules do not match: X : Foo is not included in ...(Z) = sig type t end 
  2. Module Y matches the expected module type
  3. Modules do not match: X : Foo is not included in A 
  4. Module Y matches the expected module type A
  5. The following extra argument is provided X : Foo
```
Here the application errors come from a path. To make the diff more digestible, some names have been elided (`(Y : ...)`) when unnecessary, and some names have been synthesized and defined later `...(Z)`.

Note that we use colors quite a lot to make it easier to digest:
![Functor application error with colors](https://i.imgur.com/1qlWMSu.png)

All this also works with signature inclusion checks, and the phrasing is adapted in consequence: 
```ocaml
# module F: functor (X:empty)(Y:empty) -> empty =
    functor(X:empty)(Y:empty)(Z:empty) -> Empty ;;
Error: Signature mismatch:
       Modules do not match:
         functor (X : empty) (Y : empty) (Z : empty) -> ...
       is not included in
         functor (X : empty) (Y : empty)  -> ...
  1. Module types empty and empty match
  2. Module types empty and empty match
  3. An extra argument is provided of module type empty
```

Bonuses (see the new tests for examples):
- Dependent arrows are properly treated by the diffing computation, which will try to propagate the right module typing while exploring the "best" path.
- Formatting and printing for module expression is mildly improved
- Extra care was given to errors in case of type shadowing, so that it stays reasonably understandable
- On the way, @Octachron improved the treatment of errors in case of module equivalence, which now explicitly state it was testing for equivalence

# Properties

The system was design with the following properties in mind:
1. If the error suggests a set of change to an application, the suggested application will indeed typecheck.
2. If a prefix of application succeeded, the prefix should be kept (see Caveat below).
3. We only ever provides "detailed expansions" (more than one suberror) once. This avoids having nested diffs with many cases each time.
4. All errors except functors are completely unchanged.

I believe these are sufficient to make the new errors behaves in an intuitive way for users

# Limitations

1. We do not account for swaps. This limitation is due to error-reporting: we simply don't know how to represent swaps to the users. One thing that could be considered would be to compute such swaps and only show them if the error is only a permutation of the argument. This could be done in a separate PR.
2. ~~We consider the set of parameters as static. It's possible to make variadic functors (see 77caef786), in which case our analysis is less good. It can be fixed if necessary.~~ Variadic functors are now properly accounted for, and the code is even cleaner!

# Implementation

In order to be able to compute these new errors, we had to recover the complete module application. We thus largely refactored the code for functor applications in `Env` and `Typemod` to suit our needs.
We also made it cleaner and factorized by isolating the part actually doing the application. On the way, we merged several errors that used to be distinct.

The inclusion check now also returns more structured errors for signature inclusions: in case of errors, it keeps trying to compute the coercion and returns all fields that succeeded, were missing, or errored. In general, we tried to compute structured errors before turning them into a linear printable representation. It's not perfect yet but it could be useful for rich editors.

Finally, the diffing computation is done by a classical eager diff algorithm to compute [edit distances](https://en.wikipedia.org/wiki/Edit_distance) with Insertion, Deletions and Changes. More concretely, the algorithm computes a matrix of inclusion checks and selects the less erroneous path. The implementation is *not optimized* on purpose, to stay relatively simple. In our observation, the perfs are more than sufficient for errors. The implementation is very (too?) generic, and could be reused for other diffs, if need be.

# Future work

This PR is a first step before trying to apply diffing to more errors. In particular, all is now in place to give very precise diffs for signature inclusions, which is our next goal. It's also possible to apply all this to big structural types (objects, polymorphic variants) and even function applications.
There are also lot's of refinement that could be made to these new errors (refining the metrics for "error distances", improving the phrasing, better compute "local aliases" for large types). We think all those should be made in subsequent PRs.

Drup committed:
Add module to compute edit distance generically.

Drup committed:
Expose printers for module exprs.

Drup committed:
Add tag for direct styled output.

Drup committed:
Fix format boxes in Includecode.

Drup committed:
Add printers for functor parameters without a returned signature.

Drup committed:
Improve error messages for functor signatures and applications.

Also improve module type equality errors.

Drup committed:
Add tests for functor signature and parameter errors.

Octachron committed:
add an example of variadic functor

Drup committed:
Changes for functor errors.

Drup commented:
So, any opinions (or reviews) ? :)

gasche commented:
@lpw25 any chance you would be interested in reviewing this?

lpw25 commented:
I'd be happy to review it, but I'm on holiday right now and it may be a little while before I can find the time.

garrigue commented on testsuite/tests/typing-modules/applicative_functor_type.ml:
```diff
@@ -19,13 +19,17 @@ type t = Set.Make(M).t
+       These arguments:
+         M
+       do not match these parameters:
+         functor (Ord : Stdlib__set.OrderedType) -> ...
```
Why is `Set` expanded here?

Octachron replied:
This is a regression that I missed, `Stdlib__set` is not compressed to `Set` because it is not printed in the right environment. I will fix this regression.

garrigue commented on testsuite/tests/typing-modules/applicative_functor_type.ml:
```diff
@@ -43,15 +47,19 @@ type t = F(M).t
+     Modules do not match:
+       M : sig type t = M.t val equal : 'a -> 'a -> bool end
+     is not included in
+       ...(X) = sig type t = M.t val equal : unit end
```
I do not like this `...(X)` notation very much.
Couldn't we just reuse the name of the argument when available (or synthesize one)?
```ocaml
X : sig type t = M.t val equal : unit end
```

Drup replied:
That's exactly what happens: we reuse the name if we can, otherwise, if we need to name a complex signature, we synthesize a name, often by using the parameter.

We went through various ways of naming these signatures, but none of them felt exactly right, as we need to be very clear about the fact that this name doesn't exist in the code (otherwise, the error message is very confusing). `...(NAME)` is the best we came up with.

lpw25 replied:
I agree that this notation is not good. Simply using a name or number with a sigil to indicate it is fabricated would be better e.g. `$name` or `$1`.

Drup replied:
We tried `$number`, which was mostly unreadable. We can switch back to `$name` while keeping the same way of choosing names. I'm not very convinced it's much better, but at this point I think we need to try in the wild and gather feedback.

Drup replied:
We have now switched to `$name` and remove cases where it introduces inconsistencies. I think further improvement/refinements should wait until we have more feedback.

garrigue commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1189 @@
+22 |   end
+23 | end
+Error: Signature mismatch:
+       Modules do not match:
```
Do you really need to repeat all the steps here?

Octachron replied:
I intend to remove this behavior, but this is the current behavior; and I wanted to avoid changing too many parameters at the same time.

garrigue reviewed (commented):
I didn't look at the details of the code of this huge patch, but it looks like a big step in the right direction.
Namely, module/functor application errors can be hard to debug, and this patch provides relevant information in a relatively compact format.
To provide a bit of context, the current error reporting was the result of attempts at making shorter and more relevant error messages, after a report of MB sized error outputs with the old approach. So it was mostly about filtering too big parts out, replacing them by logical positions. This patch adds to that extra information that can actually help fixing errors.
The test file demonstrates a few pathological cases. It could be good to improve the output in such cases, for instance for deeply nested modules.

Octachron committed:
update dune file

Octachron committed:
Split the application and inclusion update

This is nearly necessary to avoid mixing different concerns
and concepts together

Octachron committed:
Edit_distance with dynamical resizing

Co-authored-by: Gabriel Radanne <drupyog@zoho.com>

Octachron committed:
Pretty error message for variadic functors

Octachron committed:
More test for exotic functors

Octachron committed:
Compute list elements earlier with Includemod.dlist

Octachron committed:
Pretty error message for variadic functor inclusion

Octachron commented:
After a discussion with @garrigue , we have extended this PRs to handle correctly all functors.

All users of variadic functors,

```ocaml
module F(X:sig type witness module type t module M:t end) = X.M
module PF = struct
  type witness
  module type t = module type of F
  module M = F
end
module W = F(PF)(PF)(PF)(PF)(PF)(F)
```
can now enjoy the improved error messages too:

```ocaml
Line 1, characters 11-35:
1 | module W = F(PF)(PF)(PF)(PF)(PF)(F)
               ^^^^^^^^^^^^^^^^^^^^^^^^
Error: The functor application is ill-typed.
       These arguments:
         PF PF PF PF PF F
       do not match these parameters:
         functor (X : ...) (X : ...) (X : ...) (X : ...) (X : ...)
         (X : ...(X)) -> ...
  1. Module PF matches the expected module type
  2. Module PF matches the expected module type
  3. Module PF matches the expected module type
  4. Module PF matches the expected module type
  5. Module PF matches the expected module type
  6. Modules do not match:
       F :
       functor (X : sig type witness module type t module M : t end) -> X.t
     is not included in
       ...(X) = sig type witness module type t module M : t end
     Modules do not match:
       functor (X : ...(X)) -> ...
     is not included in
       functor  -> ...
        An extra argument is provided of module type
            ...(X) = sig type witness module type t module M : t end
````

This is done by adapting our edit distance computation to allow dynamical resizing of
the number of rows and columns.
The new commits also split the code for the inclusion and application path to avoid mixing together incompatible concepts and concerns.
Overall the code should be cleaner, and this PR should be feature complete.

gasche commented:
During a recent development meeting @lpw25 volunteered to review this.

Octachron committed:
Improved functor errors with extra arguments

Handle Map.Make(Ord)(X)(Y)(Z) and friends

Drup commented:
This was presented at the ML workshop by @Octachron ! The video is available here: https://www.youtube.com/watch?v=n3eS-yuSJ_U

Octachron committed:
Fix missing cmi case

Make the handling of abstract module types clearer

Octachron commented:
I have rebased this PR on trunk.

This rebasing comes with two bonus commits: the first new commit fixes the handling of missing cmis. The second one improves the error message for the case `Map.Make(String)(String)`. (The improved error message was presented in the ML workshop talk above).

Since this PR moves around large blocks of codes, this rebasing have required a non-trivial amount of work, and I would like to avoid repeating this across too many OCaml versions.

@lpw25 , what is the current frequency of the review ping?

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+(**************************************************************************)
```
Why the compilerlibs prefix? We don't do that for any other files.

Octachron replied:
To avoid potential conflicts since it is a new modules in compilerlibs with an otherwise quite generic name.

lpw25 replied:
It doesn't seem any more generic than some other compiler-libs modules. I think we should either:

1) Prefix everything, possibly using some dune-like namespacing or packing
2) Prefix nothing, as we currently do
3) Only prefix things when we know of a specific conflict

The approach of "randomly prefix some files some of the time" seems worse than all of those.

lpw25 commented on utils/compilerlibs_edit_distance.mli:
```diff
@@ -0,0 +1,53 @@
+  update:(('a, 'b, 'c, 'd) change -> 'state -> 'state) ->
+  'state -> 'a array -> 'b array -> ('a, 'b, 'c, 'd) patch
+
+
```
The mention of lines and columns in the comment and the record is not obvious. The rest of the interface talks of only `'a` and `'b`. Allowing arbitrary mutations to the internal state of the algorithm does not seem like a clean interface and it is not clear which mutations are valid.

If you are going to have the inputs dynamically increase in size based on the state then why ask for arrays as input in the first place?

Octachron replied:
The `diff` function is the simpler non-dynamically increasing variant, where it does make sense to have input arrays.

lpw25 commented on utils/compilerlibs_edit_distance.mli:
```diff
@@ -0,0 +1,53 @@
+(**************************************************************************)
+
+
+type ('a, 'b, 'c, 'd) change =
```
Not immediately obvious what the type parameters represent, either a comment or some names on the fields would make the code easier to read.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+    col: 'col array;
+    inner:'inner
+  }
+
```
Presumably this is a previously existing edit distance algorithm. Mentioning which one and linking to a decent description would be useful.

Octachron replied:
Good point. I will improve the documentation.

Drup replied:
(And just for clarity before we write the docs: it's a very slightly modified, non-optimized [Wagner Fischer](https://en.wikipedia.org/wiki/Wagner%E2%80%93Fischer_algorithm), which is a classical diffing algorithm. We took a simple well known algorithm on purpose)

lpw25 commented on testsuite/tests/typing-modules-bugs/pr7414_2_bad.compilers.reference:
```diff
@@ -1,20 +1,23 @@
-File "pr7414_2_bad.ml", line 46, characters 28-34:
+File "pr7414_2_bad.ml", line 46, characters 22-35:
 46 |   let module Ignore = Force(Choose) in
-                                 ^^^^^^
```
I think the new error message is strictly worse in this case. It adds a load of extraneous detail for no real gain. For functor applications where only a single argument is wrong it is simpler for the error to point directly to that argument and skip talking about the function application in general. A simple message along the lines of "This module has type ... a module was expected with type ..." would be nice and clear.

Octachron replied:
Specializing the error message for an edot distance of one sounds like a great idea.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+       Modules do not match:
+         functor (X : empty) (Y : empty) (Z : empty) -> ...
+       is not included in
+         functor (X : empty) (Y : empty)  -> ...
```
Could we get an `f = ` here to make it clearer where this expected type has come from?

Octachron replied:
Tracking that information seemed quite involved.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+         functor () (X : ...(X)) -> ...
+  1. Module () matches the expected module type
+  2. the functor was expected to be applicative at this position
+|}]
```
s/`the`/`The`/

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+         functor  (X : ...(X)) -> ...
+       is not included in
+         functor (X : ...(X)) (X : ...(X)) -> ...
+  1. An argument appears to be missing with module type
```
This error seems to use `...(X)` to refer to two different module types.

Octachron replied:
This was fixed by the switch to synthetic names of the form $T1, $T2.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+         X  Z
+       do not match these parameters:
+         functor (X : x) (Y : y) (Z : z) -> ...
+  1. Module X matches the expected module type x
```
The alignment of the list of arguments seems off. I would have expected it to line up with the "do not match ..." line.

Octachron replied:
This was fixed, in a quite indirect way. The issue is that the box structure of the main message is not shared with submessages, it is thus impossible to recover the right alignment using the box structure.

However, using tabulation boxes works.

For future works, it might be interesting to look a bit more closely at the structure of error submessages however.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+       P.B : b/1
+     is not included in
+       b/2
+      Line 5, characters 2-15:
```
Alignment of `Line 5` seems off by one. I would expect it to line up with `is not included ...` above.

Octachron replied:
Ghost white spaces have been eliminated.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+          ...(B) = sig type yb end
+        does not include
+          ...(B) = sig type ybb end
+        The type `yb' is required but not provided
```
This is the line that you are trying to highiight, since it is the most precise piece of error information that you are providing. However, it is being buried in the middle of the error message because of the argument after it. I think it would we worth considering displaying the nested information after all arguments, so that the most useful information is always at the bottom of the error message.

Octachron replied:
I am not fond of the idea of interspersing different error messages and ending with something akin to
```
1
2
... See below
3
2.1
... See below
2.2
2.3
2.1.1
```
Preserving the branching structure seems more readable.

An option might to reverse the list of the parameters, to have the last error message corresponds to the first argument where things went wrong. But that doesn't seem ideal either.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+       A.X : A.Honorificabilitudinitatibus
+     is not included in
+       A.A
+
```
Why the blank line?

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+       ...(R) = sig type upon type the type heath end
+     The type `heart' is required but not provided
+|}]
+
```
The above message seems to contradict property 3 from the description of the feature:
> We only ever provides "detailed expansions" (more than one suberror) once. This avoids having nested diffs with many cases each time.

Octachron replied:
Indeed, this part of the message was not deemed as an expansion. This has been fixed.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+         end
+       In module F:
+       Modules do not match:
+         functor (Wrong : ...(Wrong)) (X : ...(X)) X/2.T X/2.T -> ...
```
The same module is referred to as both `X` and `X/2`.

Octachron replied:
This was fixed, by disabling the disambiguation of module names in functor argument. This avoids the need to track precisely which identifiers is canonically identified to `X` during the printing of functor arguments. I would prefer to delay such tracking to further work.

lpw25 commented on testsuite/tests/typing-modules/functors.ml:
```diff
@@ -0,0 +1,1636 @@
+|}]
+
+
+(** The price of Gluttony: glutton update of environment leads to a non-optimal edit distance. *)
```
I don't think you can use glutton as an adjective, should probably be "gready".

lpw25 commented on typing/env.ml:
```diff
@@ -638,14 +638,15 @@ let components_of_module_maker' =
 
 let components_of_functor_appl' =
-  ref ((fun ~loc:_ _f _env _p1 _p2 -> assert false) :
-          loc:Location.t -> functor_components -> t ->
```
Why the tuple? Seems to just add an allocation for no gain.

Drup replied:
Because a function `... -> Path.t -> functor_components -> Path.t -> ....` causes lot's of swearing-induced debugging because I constantly swapped the two paths. The tuples makes it very clear which path is associated to the functor component and avoid all issues.
I tested with labels, it's strictly less clear and more verbose.

Finally, I think we can agree than a single short-lived tuple allocation is pretty negligeble compared to "compute the components of this functor application" and is a very low cost for absence of horrible to debug errors.

lpw25 replied:
I find the tuple adds confusion when I read the code. If you want to avoid the risk of conflating the different `Path.t` parameters then please use labels instead.

Drup replied:
This is now fixed.

lpw25 commented on typing/includemod.ml:
```diff
@@ -45,10 +45,83 @@ type pos =
-exception Error of error list
-exception Apply_error of Location.t * Path.t * Path.t * error list
+
+module E = struct
```
Module name too short and missing indentation on the contents.

lpw25 commented on typing/includemod.ml:
```diff
@@ -18,6 +18,7 @@
 open Types
+module Diff = Compilerlibs_edit_distance
 
 type symptom =
```
This seems to be unused now

lpw25 commented on typing/includemod.ml:
```diff
@@ -45,10 +45,83 @@ type pos =
+
+end
+
+type error = pos list * Env.t * symptom
```
This seems to be unused now

lpw25 commented on typing/includemod.ml:
```diff
@@ -93,72 +166,75 @@ let type_declarations ~loc env ~mark ?old_env:_ cxt subst id decl1 decl2 =
-let expand_module_alias env cxt path =
-  try (Env.find_module path env).md_type
+let expand_module_alias env path =
+  try Ok ((Env.find_module path env).md_type)
```
Nitpick: slightly cleaner using `match`

lpw25 commented on typing/includemod.ml:
```diff
@@ -253,76 +329,137 @@ let simplify_structure_coercion cc id_pos_list =
+  | Mty_functor (p, res) -> retrieve_functor_params (p :: before) env res
+  | Mty_signature _ as res -> List.rev before, res in
+  retrieve_functor_params [] env x
+
```
Style: Indentation and line breaking is a bit all over the place here.

lpw25 commented on typing/includemod.ml:
```diff
@@ -253,76 +329,137 @@ let simplify_structure_coercion cc id_pos_list =
   then Tcoerce_none
   else Tcoerce_structure (cc, id_pos_list)
 
+let retrieve_functor_params env x =
```
`x` isn't generic here so some kind of name is warranted. `mty` would be the usual choice.

lpw25 commented on typing/includemod.ml:
```diff
@@ -460,53 +592,75 @@ and signatures ~loc env ~mark cxt subst sig1 sig2 =
-        value_descriptions ~loc env ~mark cxt subst id1 valdecl1 valdecl2
+  | [] -> [], []
+  | (sigi1, sigi2, pos) :: rem ->
+      let id, item, keep = match sigi1, sigi2 with
```
Maybe `runtime` or similar instead of `keep`? Might make it clearer why some things are kept and others aren't.

lpw25 commented on typing/env.ml:
```diff
@@ -2456,6 +2464,37 @@ and lookup_functor_components ~errors ~use ~loc lid env =
+  let lid_f0, args0 = decompose_apply lid0 in
+  let path_f0, comp_f0 =
+    lookup_module_components ~errors ~use ~loc lid_f0 env in
+  let check_one_apply ~errors ~use ~loc (lid,comp) lid_arg env =
```
Why the tuple?

Drup replied:
To group together all the information regarding the functor to apply. In practice, that API is much less error prone than the previous one where it's a mixmatch of longidents and other information.

lpw25 replied:
For me tuple parameters are a code smell. If the grouping of these components were natural then they would be used for more than just the call, at which point you could give their product a proper record type. If they are just vaguely related then labelled parameters are a superior mechanism for making clear the role of each parameter since it uses actual names instead of just positions.

lpw25 commented on typing/env.ml:
```diff
@@ -2456,6 +2464,37 @@ and lookup_functor_components ~errors ~use ~loc lid env =
+        let comps =
+          !components_of_functor_appl' ~loc (path, comp_f) path_arg env in
+        let path = Papply (path, path_arg) in
+        let lid = Lapply (lid, lid_arg) in
```
Can we avoid creating artificial longidents?

lpw25 commented on typing/env.ml:
```diff
@@ -2456,6 +2464,37 @@ and lookup_functor_components ~errors ~use ~loc lid env =
+    let comp_f, mty_param = get_functor_components ~errors ~loc lid env comp in
+    let path_arg, md_arg = lookup_module ~errors ~use ~loc lid_arg env in
+    check_functor_appl
+      ~errors ~loc (lid0, path_f0, args0)
```
I don't love that we are passing the original arguments list as long identifiers. This means that we are looking up some of the arguments twice, and I generally think we should aim to throw away long identifiers as soon as they are looked up. This way we avoid the risk of repeating warnings or of accidentally using a different environment and getting inconsistent results.

Since we are clearly checking applications as a whole now, maybe we should lookup all the arguments before checking that the applications are well typed. This probably gives better error messages anyway, and we can keep paths and module types for the arguments list instead of long identifiers.

Drup replied:
I agree. We wanted to avoid changing the order of typechecking in the normal path, but we could do that.

lpw25 commented on typing/includemod.mli:
```diff
@@ -87,7 +88,16 @@ type pos =
   | Body of functor_parameter
 type error = pos list * Env.t * symptom
```
Seems to be unused now

lpw25 reviewed (commented):
I've reviewed the tests, the interfaces and most of the "glue" code. It remains for me to review the core diffing logic in `includemod.ml`, the edit distance algorithm and the changes to `typemod.ml`.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+    for j = 0 to tbl.cols do max_at tbl.lines j done;
+    !lines, !cols
+
+  let make ~size =
```
`size` seems a slightly misleading name since a matrix of "size" `(0, 0)` is not empty.

Also two labelled arguments would be clearer than a single labelled tuple that is immediately destructed. Alternatively, a `shape` record used by `create`, `real_shape` and `reshape` would probably be even better.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+  | [] -> None
+  | (m,res)::l ->Some(aux m res l)
+
+let update_cell m i j ?p ~x ~s =
```
Are `x`, `s` and `p` standard terms? If not then a full word label would be much clearer.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+    inner:'inner
+  }
+
+module Matrix = struct
```
You have a separate module here and some functions for accessing and manipulating the matrices, but you don't make the type abstract and some of the later code just reaches into the type and accesses things directly. I think it would be clearer if you put a signature on this module and abstracted the type. That would make clear what the interface to these matrices actually is.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+  in
+  update_cell tbl i j ~x:newweight ~s:newstate ~p:newres
+
+let compute_matrix ~weight ~test ~update state0 =
```
Nitpick: the parameters are in a different order here than in `compute_inner_cell` for no obvious reason.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+
+
+let select_best_proposition l =
+  let rec aux m0 res0 = function
```
`aux` appears to just be a fold.

The whole of `select_best_proposition` can also be expressed as a single fold and would probably be clearer that way.

lpw25 commented on utils/compilerlibs_edit_distance.ml:
```diff
@@ -0,0 +1,213 @@
+  let compute_proposition i j diff =
+    let* diff = diff in
+    let+ state = tbl.state.(i).(j) in
+    weight diff + tbl.Matrix.weight.(i).(j), (diff, update diff state)
```
Appears to be computing the new state eagerly even though it might not be used.

lpw25 reviewed (changes requested):
I've made a pass at reviewing the edit distance code, and I think there are a couple of fundamental issues that need to be addressed. Both issues are all related to the existence of dependence between functor parameters, something which does not exist when calculating the edit distance for ordinary strings.

#### "Path dependence" in patches
The first issue is that the dependence between parameters means that the optimum choice of patch for later parameters depends on the specific choice of patch for earlier parameters. This means that choosing a patch with a greater distance for earlier parameters can produce a small edit distance overall. The Wagner-Fischer algorithm explicitly assumes that there is no such dependence, which means that we do not always get the patch with the smallest edit distance. For example:
```ocaml
module type A = sig type a end
module A = struct type a end
module AT = struct module type T = A end

module type B = sig type b end
module B = struct type b end
module BT = struct module type T = B end

module type Ty = sig module type T end

module Foo (X : Ty) (_ : X.T) (_ : X.T) (_ : X.T) (_ : X.T) (_ : X.T) = struct
  type t
end

type fine = Foo(BT)(B)(B)(B)(B)(B).t
type broken = Foo(AT)(BT)(B)(B)(B)(B)(B).t
[%%expect {|
module type A = sig type a end
module A : sig type a end
module AT : sig module type T = A end
module type B = sig type b end
module B : sig type b end
module BT : sig module type T = B end
module type Ty = sig module type T end
module Foo :
  functor (X : Ty) -> X.T -> X.T -> X.T -> X.T -> X.T -> sig type t end
type fine = Foo(BT)(B)(B)(B)(B)(B).t
Line 16, characters 14-42:
16 | type broken = Foo(AT)(BT)(B)(B)(B)(B)(B).t
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: The functor application Foo(AT)(BT)(B)(B)(B)(B)(B) is ill-typed.
       These arguments:
         AT BT B B B B B
       do not match these parameters:
         functor (X : Ty) X/2.T X/2.T X/2.T X/2.T X/2.T  -> ...
  1. Module AT matches the expected module type Ty
  2. Modules do not match:
       BT : sig module type T = B end
     is not included in
       X/2.T
     The type `a' is required but not provided
  3. Modules do not match:
       B : sig type b = B.b end
     is not included in
       X/2.T
     The type `a' is required but not provided
  4. Modules do not match:
       B : sig type b = B.b end
     is not included in
       X/2.T
     The type `a' is required but not provided
  5. Modules do not match:
       B : sig type b = B.b end
     is not included in
       X/2.T
     The type `a' is required but not provided
  6. Modules do not match:
       B : sig type b = B.b end
     is not included in
       X/2.T
     The type `a' is required but not provided
  7. The following extra argument is provided B : sig type b = B.b end
|}]
```
Here we get a proposed patch with 6 changes instead of the simplest fix which only requires 1 change.

I think that this is broadly fine -- we do not need to guarantee the patch with the smallest edit distance. However it shows that the interaction with parameter dependence is subtle and means that there is not currently any proper specification of what the edit distance code is actually doing. The specification can't just be "compute the patch with the smallest edit distance" since it doesn't do that, so I think some other specification needs to be provided.

#### Dynamic expansion still treats possible paths as a matrix
The dynamic version has a number of issues. The interface is extremely under-specified: it just directly exposes the underlying state of the algorithm, allowing the user to edit it without any guidelines as to what modifications are valid.

Fundamentally, it is trying to handle the fact that different choices of patch can produce input "strings" of different lengths. This fact breaks the basic assumption that the possible choices of patch form a matrix. However, the current implementation tries to gloss over this and keep the matrix representation. The result is that only patches of the longest available length can be selected, for example:
```ocaml
module type A = sig type a end
module A = struct type a end
module type B = sig type b end
module B = struct type b end

module type ty = sig type t end

module type Ext = sig module type T module X : T end

module AExt = struct module type T = A module X = A end
module FiveArgsExt = struct
  module type T = ty -> ty -> ty -> ty -> ty -> sig end
  module X : T =
    functor (_ : ty) (_ : ty) (_ : ty) (_ : ty) (_ : ty) -> struct end
end

module Bar (W : A) (X : Ext) (Y : B) (Z : Ext) = Z.X

type fine = Bar(A)(FiveArgsExt)(B)(AExt).a
type broken = Bar(B)(FiveArgsExt)(B)(AExt).a
[%%expect {|
module type A = sig type a end
module A : sig type a end
module type B = sig type b end
module B : sig type b end
module type ty = sig type t end
module type Ext = sig module type T module X : T end
module AExt : sig module type T = A module X = A end
module FiveArgsExt :
  sig module type T = ty -> ty -> ty -> ty -> ty -> sig end module X : T end
module Bar : functor (W : A) (X : Ext) (Y : B) (Z : Ext) -> Z.T
type fine = Bar(A)(FiveArgsExt)(B)(AExt).a
Line 20, characters 14-44:
20 | type broken = Bar(B)(FiveArgsExt)(B)(AExt).a
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: The functor application Bar(B)(FiveArgsExt)(B)(AExt) is ill-typed.
       These arguments:
           B FiveArgsExt B AExt
       do not match these parameters:
         functor (W : A) (X : Ext) (Y : B) (Z : Ext) ty ty -> ...
  1. An argument appears to be missing with module type A
  2. An argument appears to be missing with module type Ext
  3. Module B matches the expected module type B
  4. Module FiveArgsExt matches the expected module type Ext
  5. Modules do not match:
       B : sig type b = B.b end
     is not included in
       ty
     The type `t' is required but not provided
  6. Modules do not match:
       AExt : sig module type T = A module X = A end
     is not included in
       ty
     The type `t' is required but not provided
|}]
```
Here the algorithm ignores the simplest patch with only a single change in favour of the longer patch which has 4 changes.

Even if this particular bug is fixed the general approach seems fragile and hard to reason about. I suggest that the interface and algorithm are rewritten to more explicitly account for the dependence between parameters. I suspect that using an array for the fixed input and something more coinductive for the extensible input will be cleaner. Similarly, I think moving away from a matrix for the internal representation should make the cases where some paths extend the input easier to reason about.

gasche commented:
I was impressed with @lpw25's very nice review and curious about his question on the specification of the edit-distance in present of signature dependencies -- so I just had a chat with @Octachron about it and I'm writing down the summary here.

In the signature of the functor, some arguments are "dependent" in the sense that the signature of following arguments depends on the module passed for this argument. If we think of the standard dynamic algorithm for edit distance, we end up with "dependent columns" in the "matrix" (not really a matrix anymore), where the structure of the search space for later positions depends on the choice of letter place in the "dependent column". If we wanted to preserve optimality of the computation (shortest edit distance), we would need to try to instantiate those columns with all possible arguments, to fully explore the corresponding search space. So if we are looking at N functor arguments, without a dependent column the algorithm is O(N^2), but with just *one* dependent column there are N dependent instantiations choices, so we are in O(N^3). In general the complexity is of the form O(N^(2+D)), where D is the number of dependent columns. We (they) can implement this, it gives a nice specification, but (1) it is more complex to implement than the current approach and (2) the computational behavior is terrible in presence of dependent columns (but then nobody uses dependent columns in practice, so this is just for the beauty of science that we are discussing this question in the first place).

If you want to reduce the complexity, you have to adopt a partial strategy where, on dependent columns, only a subset of possible instantiations are explored. The current implementation is such a strategy, where exactly one instantiation is explored: one that corresponds to a shortest-edit-distance path for the prefix of the arguments before the dependent position. Intuitively this approach makes sense: we use the (greedy) heuristic that the best path so far *probably* corresponds to "what the user should have written", and thus a reasonable choice of instantiation to explore. Note that there *is* a natural specification of the result: it is not optimal, but it consists of a series of optimal choices in sub-spaces without dependent columns, with each dependent instantiation made according to this local notion of optimality.

lpw25 commented:
I think that is a good summary of my first point, and I think that:

> Note that there is a natural specification of the result: it is not optimal, but it consists of a series of optimal choices in sub-spaces without dependent columns, with each dependent instantiation made according to this local notion of optimality.

essentially addresses it. A clear comment that describes that specification in the code is all that needs to happen there.

Just for clarity: my second point still stands and is really just a bug in the current implementation. The second example I gave has an answer with only 1 change according to the "series of optimal choices in sub-spaces" specification but the algorithm instead returns an answer with 4 changes. I think this bug needs to be fixed, and I also think the implementation needs to be adjusted so that it is clear no similar bugs are still present because it is very hard to reason about it right now.

Octachron commented:
Thanks for the review! I will have a look at a more principled implementation of the dynamic case to correct this variadic bug and write down a precise specification.

Drup committed:
Make the code a bit cleaner in Includemod.

Drup committed:
Use $name as generate name in diffing errors.

Drup committed:
Capitalize error messages when appropriate.

Drup committed:
Rename Compilerlibs_edit_distance into Diffing.

Drup committed:
Formalize the internal API for Diffing.Matrix.

Drup committed:
Simplifies the external API for Diffing.

Drup committed:
First shot at documenting Diffing.

Drup committed:
Put more constraints on expansions during variadic diff.

Drup committed:
Fix a bug in the rewrite of select_best_proposition.

Drup committed:
Code simplification and debugging help in Diffing.

Drup committed:
Fix selection of final state in the diff matrix.

Drup committed:
Prefer diagonales when diffing.

Drup committed:
Add Leo's example.

Drup committed:
More code simplification.

Drup committed:
Use more labels in Env.

Drup committed:
Lookup all arguments of a functor prior to application in Env.

Drup committed:
Fix some english.

Drup commented:
While in the process of addressing @lpw25 's remarks (almost everything is fixed by now), I wondered if we could just implement the `O(N^(2+D))` algorithm mentioned by @gasche .

The pure brute force algorithm is clearly `O(2^N)`, which is almost certainly unacceptable even for error messages. but the one which only explodes on dependent columns might be okay. So far, I've seen two solutions:
1. Keep the dynamic programming nature of the current algorithm but makes the cells sets of states. This assume that we can compare "module typing states" (an env, subst and module type) for equality. I'm not sure how to do that without many more module type equivalence tests.
2. Make an hybrid algorithm which forks inductively on dependent columns, but uses local dynamic programming with the matrix on non-dependent columns. However this requires discerning dynamic columns, which is equivalent of knowing if a subst leaves a given module type unchanged. I do not know how to do that reasonably efficiently in the current typechecker.

In any case, I doubt it's worth the complexity, although it would indeed make for a simpler spec.

Drup committed:
Check typo.

gasche commented:
My personal impression is that it would be pragmatic to go ahead with documenting the current specification and hope for Leo's approval as a reviewer. The corner-case of dependent columns is interesting to think about, but for now we have no reason to believe that it makes any difference for actual users. If we learn later that it does, you can always revisit (and the discussion here would then be very useful).

Drup commented:
Oh yes, it was mostly an academic (I mean, even more than usual) discussion. The plan is still to use the current approximation.

Drup committed:
Attempt at a clearer description.

Drup committed:
Some internal comments.

Octachron committed:
Shorter messages for one argument diff

Octachron committed:
Fix empty line in error messages

Octachron committed:
Control alignment with tboxes

Octachron committed:
Fix misaligned conflict explanation

Octachron committed:
Suberrors in signature require an expansion token

Octachron committed:
Check typo

Octachron committed:
diffing: avoid copy in the non-extension case

Octachron committed:
diffing: more explict name for update kind

Octachron committed:
diffing: more descriptions and some renaming

Drup committed:
Make the selection of final cell more principled.

Drup committed:
Fix comment.

Octachron committed:
diffing: update printing env for functor params

Octachron committed:
copyright headers

Octachron committed:
diffing: update documentation

Octachron committed:
diffing: coherent functor argument printing

Octachron committed:
diffing: add parentheses in functor (sig end)

Octachron commented:
I have pushed an updated version of the PR with an cleaner API for the diffing module and a lot of small printing fixes.

The `Diffing` interface now details that it returns optimal patch only when the optimal patch does not depend on the inner patch. When this properties is not verified, we return a correct patch, whose subpatchs are optimal on the non-dependent sections.

The new API now only allow extensions in one direction, without access to the inner state.

The algorithm is still using a matrix. It was a simple way to express the fact that we are classifying patches between k-prefix and l-prefix while using a heuristic to choose a "best" representative among all k-to-l patches. We could switch to a more graph-oriented graph-oriented implementation, but since the matrix will have to stay for storing the best k-to-l patch this seemed suboptimal. Nevertheless, the matrix flood filling has been considerably streamlined.

The bug reported by @lpw25 in presence of multiple possible views of the lengths of the string has been fixed by considering all complete patches (patches that completely transform the input string into the expected output string) before selecting the best patch.

Octachron committed:
ci: update .depend

Drup commented:
@lpw25 Do you think you'll have the time/availability to do another round of review ? Otherwise, since you agreed with the general principle, maybe we can ask another reviewer for the second round (@trefis ?)

(Random though: Maybe we should have a system similar to journal reviews for patches considered for inclusions ...)

gasche commented:
> (Random though: Maybe we should have a system similar to journal reviews for patches considered for inclusions ...)

There are indeed strong similarities, especially the part that goes like "one year in, still waiting for my reviews...".

lpw25 commented on testsuite/tests/typing-modules-bugs/pr7414_2_bad.compilers.reference:
```diff
@@ -1,20 +1,17 @@
+Error: Modules do not match:
+       Choose :
+       functor () -> sig module Choice : T val r : '_weak1 list ref ref end
+     is not included in $T1 = functor () -> S
```
This message is still a bit worse than the old one:
- It highlights the location of the whole application rather than just the argument.
- It defines a mysterious `$T1` that has not previously appeared, which is confusing.
- The indentation seems off.

Octachron replied:
This should be completely fixed now.

lpw25 reviewed (commented):
I've looked back over the tests, glue code and new diffing code and they all look good to me. I still needed to look at the new code in env, typemod and includemod. I'm away for a couple of weeks but I should be able to find a spare few hours to do that in January.

lpw25 commented on typing/typemod.ml:
```diff
@@ -1876,6 +1877,11 @@ let wrap_constraint env mark arg mty explicit =
 
+let rec extract_application sargs smod = match smod.pmod_desc with
+  | Pmod_apply(f, arg) ->
+      extract_application ((smod, f, arg)::sargs) f
```
I think a record type might be good for the arguments of the application, and I think you only need the location and the attributes from `smod`, not the whole thing.

lpw25 commented on typing/typemod.ml:
```diff
@@ -2081,6 +2015,101 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+and type_application loc strengthen funct_body env sfunct sargs =
+  let type_arg (floc, attrs, sarg) =
+    let arg = type_module true funct_body None env sarg in
+    floc, attrs, sarg, arg, path_of_module arg in
```
Another place where a record type would be welcome. It would also make sense to merge this `type_arg` function with the `extract_application` function that generates the list.

Octachron replied:
I have added a record in this case, and inlined the previous case.

lpw25 commented on typing/typemod.ml:
```diff
@@ -2795,117 +2824,125 @@ let package_units initial_env objfiles cmifile modulename =
 
 (* Error report *)
 
+
```
I don't mind the extra space, but the comment should clearly go after it.

lpw25 commented on typing/includemod.ml:
```diff
@@ -45,10 +44,80 @@ type pos =
+    | Params of functor_params_diff
+    | Result of module_type_diff
+
+  and ('arg,'path) functor_param_syndrom =
```
s/syndrom/symptom/?

Octachron replied:
Using symptom uniformly seems indeed better.

lpw25 commented on typing/includemod.ml:
```diff
@@ -45,10 +44,80 @@ type pos =
+    | Incompatible_params of 'arg * functor_parameter
+    | Mismatch of 'path option * Ident.t option * module_type_diff
+
+  and arg_functor_param_syndrom =
```
s/syndrom/symptom/?

lpw25 commented on typing/includemod.ml:
```diff
@@ -93,72 +162,75 @@ let type_declarations ~loc env ~mark ?old_env:_ cxt subst id decl1 decl2 =
 (* Inclusion between class declarations *)
 
-let class_type_declarations ~loc ~old_env:_ env cxt subst id decl1 decl2 =
+let class_type_declarations ~loc ~old_env:_ env  subst _id decl1 decl2 =
```
No need to keep old parameters you are no longer using.

lpw25 commented on typing/includemod.ml:
```diff
@@ -93,72 +162,75 @@ let type_declarations ~loc env ~mark ?old_env:_ cxt subst id decl1 decl2 =
+      Error Error.(Core(Class_type_declarations(diff decl1 decl2 reason)))
 
-let class_declarations ~old_env:_ env cxt subst id decl1 decl2 =
+let class_declarations ~old_env:_ env  subst _id decl1 decl2 =
```
No need to keep old parameters you are no longer using.

lpw25 commented on typing/includemod.ml:
```diff
@@ -93,72 +162,75 @@ let type_declarations ~loc env ~mark ?old_env:_ cxt subst id decl1 decl2 =
-    Env.find_modtype_expansion path env
-  with Not_found -> raise Dont_match
+let expand_modtype_path env path =
+   match Env.(find_modtype path env).mtd_type with
```
Why not use `Env.find_modtype_expansion` as the previous code did?

Octachron replied:
To avoid converting to exception in `Env.find_modtype_expansion` and back to option here.

lpw25 replied:
But you have to catch `Not_found` anyway. I think saving the allocation of `Some` isn't worth the less clear code.

Also, why the local open? It seems to introduce a risk of accidental capture for no benefit.

Octachron replied:
Updated to use `find_modtype_expansion`.

lpw25 commented on typing/includemod.ml:
```diff
@@ -253,76 +325,138 @@ let simplify_structure_coercion cc id_pos_list =
+  try_modtypes ~loc env ~mark dont_match subst mty1 mty2
+
+
+and try_modtypes ~loc env ~mark dont_match subst mty1 mty2 =
```
Why the cps-style `dont_match` parameter? A variant type would be clearer.

Even better than that would be to just return `(coercion, module_type_symptom) result` and always wrap the error case in the diff. The only reason that isn't trivial is the case on line 374 -- where we currently do not wrap the result of `strengthened_modtypes` using `dont_match`. I think it would be better to add an `Alias` constructor to `module_type_symptom` to handle that case, and then not print it in error messages, rather than keeping the additional complexity in this function.

Generally the error types should be a direct reflection of the error checking algorithm, and the UX related decisions about what to print should be taken in the printing functions.

Octachron replied:
I have removed the cps `dont_match` argument by returning a result uniformly and moving part of the logic to the error type.

lpw25 commented on typing/includemod.ml:
```diff
@@ -253,76 +325,138 @@ let simplify_structure_coercion cc id_pos_list =
-          (try_expand_modtype_path env p1)
-          (try_expand_modtype_path env p2)
+        begin match expand_modtype_path env p1, expand_modtype_path env p2 with
+        | Some p1, Some p2 ->
```
Should be `mty1` and `mty2`, since they are module types not paths.

lpw25 commented on typing/includemod.ml:
```diff
@@ -460,53 +589,75 @@ and signatures ~loc env ~mark cxt subst sig1 sig2 =
-        value_descriptions ~loc env ~mark cxt subst id1 valdecl1 valdecl2
+  | [] -> [], []
+  | (sigi1, sigi2, pos) :: rem ->
+      let id, item, present_at_runtime = match sigi1, sigi2 with
```
Weird styling: put a newline after the `=`.

lpw25 commented on typing/includemod.ml:
```diff
@@ -45,10 +44,80 @@ type pos =
-exception Error of error list
-exception Apply_error of Location.t * Path.t * Path.t * error list
+
+module Error = struct
```
This file is getting pretty long. Could we extract all the error types and printing code for them into a separate file? That would also allow a clear definition the interface of this part of the code.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+
+(* Error report *)
+
+module Short_name = struct
```
Not immediately obvious what this type is or what it is for, could use some documentation.

Octachron replied:
This was renamed to `With_shorthand` with some more documentation.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+module FunctorDiff = struct
+  open Diffing
+
+  type ('a,'b) data = { data:'a; metadata:'b}
```
Most generic names possible for this type and its fields. Could we find something a little more specific? Could also use some documetation.

Octachron replied:
The whole type is gone.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+      | Unit -> None in
+    { data; metadata }
+
+  let need_expansion { env ; res ; _ } = match res with
```
I would expect something called `need_foo` to return a boolean. Perhaps, something like `try_expand_param`?

Octachron replied:
Renamed.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+    | Mty_ident _ | Mty_alias _ as mty -> Some mty
+    | Mty_signature _ | Mty_functor _ -> None
+
+  let param_preprocess data =
```
Again, a very non-descriptive name. It appears to be one of the only two constructors of the `data` type, perhaps that could be reflected in the name. Perhaps that type could get its own small module containing its constructors and destructors.

Octachron replied:
The need of this preprocessing step has disappeared after some refactorization.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+    in
+    let args = Array.map arg_preprocess @@ Array.of_list args in
+    let params = Array.map param_preprocess @@ Array.of_list params in
+
```
Please avoid blank lines in the middle of functions.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+    in
+    Diffing.variadic_diff ~weight ~test ~update state param1 param2
+
+  let data_preprocess (parg,_,_,fn) =
```
This function seems unnecessary. You only use it in a couple of places and then seem to immediately destruct the results anyway.

Octachron replied:
It is gone.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+            let res = Option.map (fun res ->
+              let scope = Ctype.create_scope () in
+              let subst = Subst.add_module param arg Subst.identity in
+              Subst.modtype (Rescope scope) subst res) st.res in
```
The styling here makes the code very hard to read. Please consider writing such code as:
```ocaml
let res =
  Option.map
    (fun res ->
       let scope = Ctype.create_scope () in
       let subst = Subst.add_module param arg Subst.identity in
       Subst.modtype (Rescope scope) subst res)
    st.res
in
```

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+        st, [||]
+    | Keep (Some arg, Named (param_name, _param), _)
+    | Change (Some arg, Named (param_name, _param), _) -> begin
+        let arg' = Subst.modtype Keep st.subst arg.mty in
```
The name `arg'` is confusing because it is only the `mty` component of `arg`, so you still use `arg` in the code below. `arg'` is also only used on one side of the following match, so this line should be pushed inside it.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+    | None -> state, [||]
+    | Some (res, expansion) -> { state with res }, expansion
+
+  let arg_update d st = match data d with
```
There seem to be two parallel aspects to this module, an "arg" part and an "app" part, which do not depend on each other. I think that putting them in separate submodules, and removing the `arg_` and `app_` prefixes from their names would make things clearer.

Octachron replied:
The code has been split in three: the diffing code for inclusion, application and some printing code has been moved to the new `Includemod_errorprinter` module.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+
+  let shortname side pos =
+    match side with
+    | `Got -> Format.sprintf "$S%d" pos
```
Instead of polymorphic variants, a variant type definition with some documentation would be clearer.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+      {Short_name. name = (shortname side pos); item = mty; from=None }
+    in
+    let elide_if_app s = match ctx with
+      | `App -> `Unneeded
```
Definitely would prefer a normal variant for `ctx`.

Octachron replied:
No more polymorphic variants.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,57 +718,382 @@ let can_alias env path =
+
+  (* Simplication for printing *)
+
+  let shortname side pos =
```
Maybe move these functions into the `Short_name` module, which could be moved down here. (I'm not quite sure as I still don't really know what they are for).

Octachron replied:
The function has been moved to the `With_shorthand` module as `With_shorthand.make`.

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+          "@[<hv 2>Values do not match:@ %a@;<1 -2>is not included in@ %a@]%a%t"
+          !Oprint.out_sig_item t1
+          !Oprint.out_sig_item t2
+        show_locs (diff.got.val_loc, diff.expected.val_loc)
```
Unexpectedly unindented

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+    let tmty = Printtyp.tree_of_modtype mty in
+    Format.dprintf "%a" !Oprint.out_module_type tmty
+
+  let definition_of_functor_param x = match Short_name.functor_param x with
```
Missing newline after = here and in the following functions.

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+    | Value_descriptions diff ->
+        let t1 = Printtyp.tree_of_value_description id diff.got in
+        let t2 = Printtyp.tree_of_value_description id diff.expected in
+        Format.dprintf
```
Elsewhere I think we use fprintf for everything, why the switch to dprintf?

Octachron replied:
This makes it easier to segment the error message in multiple suberrors while still handling properly the global state in `Printtyp`.

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
-          Some (Location.error_of_printer ~loc (report_apply_error p1 p2) err)
+      | Error err -> Some (report_error err)
+      | Apply_error {loc; env; lid_app; mty_f; args} ->
+          let prepare_arg (path_arg, md_arg, mty_arg) =
```
Could we move this up into `report_apply_error` rather than have it as part of this printer registration code?

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+          FunctorDiff.arg_diff env ctx got expected
+          |> FunctorDiff.prepare_patch ~drop:false ~ctx:`Sig
+        in
+        let got = Pp.(params_diff space (got param_id) functor_param d) in
```
This line is just evil. Please avoid the name collision between the `got` in the local scope and the `got` from the local open of `Pp`.

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+        Some (proj mty,(x,mty))
+    | Diffing.Insert _ -> None
+
+  let expected proj = function
```
`proj` seems to just be `param_id`

lpw25 commented on typing/includemod.ml:
```diff
@@ -867,30 +1277,571 @@ let is_big obj =
+    | Named (Some _ as x,_) -> x
+    | Unit | Named(None,_) -> None
+
+  let got proj = function
```
Can we avoid the higher-order `proj` parameter? It's either constant `None` or `param_id`.

Octachron replied:
Removed.

lpw25 reviewed (commented):
I've reviewed the rest of the code now. I've left many comments for improvements, but they are mostly fairly small things.

I haven't gone through the actual diffing and error printing code in includemod.ml in detail yet. I'm finding it pretty hard going to understand those parts -- its full of ad-hoc code and there is no high-level description or type definitions to give a clear picture of what's going on. I think it would be good to lift that code into its own file, and to spend some effort making it easier to review.

I have reviewed the other parts of includemod.ml though, and I'm happy that this PR shouldn't introduce any bugs in the actual type-checking machinery.

Octachron commented:
Thanks for the review. Extracting and documenting the printing code (and fixing the last printing regression) is indeed a good point. I shall work on that (next week?). Note that I also plan to simplify all this printing code once we depart from the current display scheme for errors in nested modules.

Octachron committed:
review: add an application summary type

Octachron committed:
review: remode metadata type

Octachron committed:
review: white spaces

Octachron committed:
review: unused parameters

Octachron committed:
review: use symptom uniformly

Octachron committed:
review: remove dont_match

Octachron committed:
review: simplification after dont_match_removal

Octachron committed:
review: module types are not path

Octachron committed:
review: style

Octachron committed:
review: rename need_expansion

Octachron committed:
review: Option.map formatting

Octachron committed:
review: Pp.{got,expected}

Octachron committed:
review: split includemod module

Octachron committed:
review: split module with header files

Octachron committed:
review: only one new module

Octachron committed:
review: split functor diff modules

Octachron committed:
review: remove unused argument in functor diffs

Octachron committed:
review: rename arg' and move in the right branch

Octachron committed:
review: less adhoc printing for functor param errors

Octachron committed:
review: small tweaks

Octachron committed:
review: no more sub messages

Octachron committed:
review: fix previous commit

Octachron committed:
review: includemod_errorprinter small simplifications

Octachron committed:
review: the great reorganization

Octachron committed:
review: specialized error message for single argument and change

Octachron committed:
includemod_errorprinter: remove unused field

Octachron committed:
includemod_errorprinter: simplify functor arg description

Octachron committed:
review: remove polymorphic variant

Octachron committed:
includemod_errorprinter: simplify Functor_suberror

Octachron committed:
review: Short_name ⇒ With_shorthand

Octachron committed:
includemod_errorprinter: more explicit names

Octachron committed:
probable fix: remove Btype.snapshot

Octachron committed:
includemod_errorprinter: duclicate Inclusion.insert

Octachron committed:
includemod_errorprinter: better name

Octachron commented:
I have pushed a new version of the code, that separate the error printing code in its own module `Includemod_errorprinter` with an improved organization.

The diffing code itself should be also clearer: the inclusion and application part are now clearly separated, there is no more generic sounding types and functions, and the printing code has been moved to the `Inclusion_errorprinter` module. 

Separating the error definitions from `Includemod` in a natural way tended to create too many small modules. I would rather split them if it is still deemed necessary after this first split.

Octachron committed:
please depend CI

Drup commented:
@lpw25 I think we have addressed all your concerns. The code is now very clean and the CI failures seems erroneous. Do you aim to review again ? @Octachron and I always intended to keep improving this error-reporting code when working on signatures and other diffable elements. I think that, at this point, spending more time polishing the code is counterproductive and would like to have it merged, now that we have established it will not introduced any bugs.

gasche commented:
@dra27 there seems to be a small bug in our current Github Action code, observed (I suppose) on pull requests with many commits (that require exponential deepening etc.).

```
Run tools/ci/actions/check-changes-modified.sh 'pull_request' 'trunk' 'da9791ec9111fc941508764bb7d3a5167de07fcb' 'semdiff_includemod' 'b849aa15c453d93ac3a6adf450e74d041b8ac913'
Determining merge-base of da9791ec9111fc941508764bb7d3a5167de07fcb..b849aa15c453d93ac3a6adf450e74d041b8ac913 for semdiff_includemod
 - Deepening by 50 commits
Error: Process completed with exit code 128.
```

dra27 commented:
Ack - just working on it

dra27 commented:
CI check fixed in https://github.com/ocaml/ocaml/commit/3ef9ce800fea63c90294fe306e552b4b5fafaa1a

gasche commented:
Great, thanks!

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,339 +732,293 @@ let can_alias env path =
+    | Keep (_,Unit,_)
+    | Change (_,(Unit | Named (None,_)), _) ->
+        st, [||]
+    | Insert (Named (Some p, arg))
```
Would you mind changing the various `p` variables to be `id` in this function? `p` tends to be used for `Path.t` rather than `Ident.t` throughout the type-checker, so it's easy to get a bit confused here.

Octachron replied:
Done (in the whole file).

lpw25 commented on typing/includemod_errorprinter.mli:
```diff
@@ -0,0 +1,17 @@
+(**************************************************************************)
+
+val err_msgs: Includemod.explanation -> Format.formatter -> unit
+val register: unit -> unit
```
I don't think a `register` function is a good API. If you lifted the error types out of `Includemod` then you could expose a printing function here and do the registration in `includemod.ml` where the handled exceptions are defined.

Octachron replied:
Possibly, in my previous tentative, there is some helpers functions that are needed both in the printer and the typing code, and which results in few splintered modules.

lpw25 commented on typing/includemod.ml:
```diff
@@ -562,339 +732,293 @@ let can_alias env path =
+   The diffing computation uses the internal typing function
+ *)
+
+module Functor_inclusion_diff = struct
```
These seem more natural in `Includemod_errorprinter`, why can't you move them to there?

Octachron replied:
The diffing test is using some of the internal function of Includemod, we could expose those in a later refactoring.

lpw25 reviewed (approved):
A few small points, and I still haven't reviewed the diff printing code in detail, but I think this is ok to merge. I do think it would be a bit cleaner with the error types lifted out of `Includemod` -- either into their own module or into `Includemod_errorprinter`, but that can be done later.

Octachron committed:
review: use find_modtype_expansion

Octachron committed:
review: use id$n for ident names and not p$n

Octachron commented:
@lpw25: I have pushed the last fixes. Thanks again for taking your time for the review !

@gache: I have a history bikesheding question: do you think that the PR should be squashed to remove the back-and-forth or should we keep the individual commits?

gasche commented:
Congratulations! This was a lot of work for @Octachron and @Drup, and I was also impressed by @lpw25's superb reviews.

avsm commented:
The entire MirageOS community of over-functorised developers is _hugely_ appreciative of this change ;-) Thank you!

gasche commented:
A minor comment: when diffing applications that contain a mix of applicative arguments and generative arguments (`()`, typically at the end), the current code will show, for example:

```
       4. Module types () and () match
```

I find "module type" slightly improper here (`module type S = ()` is not valid), I would say: "Generative arguments () and () match". I don't know if this is an easy change to make.
